core/src/saros/session/internal/DeletionAcknowledgmentDispatcher.java

```diff
@@ -0,0 +1,90 @@
+package saros.session.internal;
```

Comment:
Any reason to for the session package instead of the concurrent package ?

Reply:
I was very unsure where to put this. I had it in the concurrent package at first. My reasoning to move it here was that  *technically* the ACKs are not a specific component dealing with concurrency issues, it is just currently the only usage.

But, as I already said, I am unsure on this. If you thing it fits better in the concurrency package, I don't have any issues with moving it there.

Reply:
This split on where to put it also stems from me being a tiny bit unhappy with the current solution.

It seems weird to me to filter activities in the concurrency handling classes and not in the `ActivityHandler`, which *handles activities*. This would also allow us to only run one filter on the host side (instead of two with the current solution, one for the document client and one for the document server).

But I couldn't get that solution to work as I am not familiar enough with the inner workings of the activity handler (i.e. how activities are routed). I tried to just add the functionality to the beginning of `handleIncomingActivities(...)` and `handleOutgoingActivities(...)`, but this lead to weird feedback loops (the deletion activities were dropped by the sender as they seemed to be passed through the outgoing method twice).

I tried to debug it but for a while but then just gave up as it isn't that much of an improvement.

Reply:
I have not written the routing, but yes the whole ActivityHandler is a bit tricky as it is coupled with the session. Taking a quick look it goes:

ActivitySequencer -> SarosSession -> ActivityHandler#handleIncoming (when Server then) -> callback (which is the session) -> may call ActivityHandler#handleOutgoing

General rule for activites. Every activity (unless it is a targeted activity) is send to all other users (the generator will never see it unless it is a targeted activity for himself). If the activity is a targeted activity it is send to the host who will "route" it.

Reply:
Seems I had similar issues with this and added a comment. Maybe it is helpful

https://github.com/saros-project/saros/blob/master/core/src/saros/session/internal/SarosSession.java#L751

Reply:
When the host user creates an activity the sequence is: Producer->Session->ActivityHandler#handleOutgoing->Session->ActivitySequencer->Session->ActivityHandler#handleIncoming->Session->ActivitySequencer

Where for every other user the sequence is:
Producer->Session->ActivityHandler#handleOutgoing->Session->ActivitySequencer

Reply:
Thanks for the info. But all of this just makes the approach of adding the filter to the activity dispatcher seem way more difficult than the current solution.

Furthermore, I am uncertain whether a shared filter for the client and server on the host instance would work.
My guess is that the current implementation only works as the "first loop" (call to incoming->outgoing) on the host side sets the client filter and the "second loop" sets the server filter.

Because otherwise I can't explain why the current implementation works but my approach for moving it into the `ActivityHandler` did not. The current implementation should run into the same issues as the other approach just simply moves the calls out of the respective client/server method into the `ActivityHandler` method that calls them. So the unification of the client and server filter on the host side is most likely the issue.

Even still, I don't think investing more time into this is the right move as it would require a lot more testing and debugging without providing *that* much of an architectural improvement.

@srossbach What is your opinion? (Also what is your opinion on the placement of the `DeletionAcknowledgementDispacher` with the current approach?)

Reply:
I am too tired to review this fully for now but:

1. Are you aware that a file activity can be a recovery ? Have not checked if this could mess something up.

2. You can play around with the STF that also have some methods where you can suspend and resume network traffic. https://github.com/saros-project/saros/blob/master/stf/src/saros/stf/server/rmi/controlbot/manipulation/INetworkManipulator.java

3. What happens if no acknowledge arrives, e.g a user got disconnected.

Reply:
> 1. Are you aware that a file activity can be a recovery ? Have not checked if this could mess something up.

Good point. I have not thought about that. Will check if it matters.

From the top of my head, the recovery stuff might be an issue when the host did not receive a deletion activity. The resulting recovery activity for the client would be a creation activity, which would not be filtered, so the recovery part itself is not the issue. But, as the client filters out all activities for the deleted resource, it would drop the checksum activities used to detect filesystem discrepancies, meaning the desync would never be detected. I think.... (I agree that it is too late for a more detailed analysis right now :wink:)
As I already said, will do more analysis/testing.

> 2. You can play around with the STF that also have some methods where you can suspend and resume network traffic. https://github.com/saros-project/saros/blob/master/stf/src/saros/stf/server/rmi/controlbot/manipulation/INetworkManipulator.java

I will have a look at it. But for that I would first have to get the STF to run. It doesn't seem to play nice with my window manager (it always gets stuck in a loop opening and closing the settings menu). I think there is a way to run it in a docker container, which should resolve the issue, but I haven't had a look at the necessary setup yet.

> 3. What happens if no acknowledge arrives, e.g a user got disconnected.

When the acknowledgment is just dropped/lost, the filter will remain until the file is re-created. But, as this should happen very rarely (without making the whole session unstable), the resulting memory overhead until the end of the session should be acceptable.

When a user leaves the session (or a project is removed from the session), the list of filers is updated accordingly. This is implemented through a [session listener](https://github.com/saros-project/saros/blob/a394fc7754ff362097b16dae0b3804d4b5d80c18/core/src/saros/concurrent/management/ResourceActivityFilter.java#L95).